### ISSUE TITLE
Wrong setting for TasksMax in Systemd service file

### ISSUE THREAD

**Issue by geotekberlin** [Icinga/icinga2#5611]
After upgrading from Icinga 2.7.0 to 2.7.1 we see again crashes

```
Caught unhandled exception.
Current time: 2017-09-23 06:42:33 +0200

Application information:
  Application version: r2.7.1-1
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: openSUSE Leap
  Platform version: 42.3
  Kernel: Linux
  Kernel version: 4.4.87-25-default
  Architecture: x86_64

Build information:
  Compiler: GNU 4.8.5
  Build host: e5e5f77d6059

Error: Function call 'fork' failed with error code 11, 'Resource temporarily unavailable'
```
Looking at the file /etc/systemd/system/multi-user.target.wants/icinga2.service we see that Icinga 2.7.1 has now added the line DefaultTasksMax=infinity but "systemctl show -p TasksMax Icinga2" still shows 512, so this line does not seem to have any effect, neither does my previous entry "TaskMax=infinity".

It appears that only the line "DefaultTasksMax=infinity" in the file "/etc/systemd/system.conf" has the desired effect and the Icinga addition does not work (enymore) with systemd 228 as supplied with OpenSuxe 42.3

**Comment by lazyfrosch** [comment]
The correct setting would be `TasksMax=infinity` in the `icinga2.service` unit file.

`DefaultTasksMax=infinity` is only for `systemd-system.conf`

As state in `systemd.resource-control(5)`

       TasksMax=N
           Specify the maximum number of tasks that may be created in the unit. This ensures that the number of tasks accounted
           for the unit (see above) stays below a specific limit. This either takes an absolute number of tasks or a percentage
           value that is taken relative to the configured maximum number of tasks on the system. If assigned the special value
           "infinity", no tasks limit is applied. This controls the "pids.max" control group attribute. For details about this
           control group attribute, see pids.txt[6].

           Implies "TasksAccounting=true". The system default for this setting may be controlled with DefaultTasksMax= in
           systemd-system.conf(5).

**Comment by lazyfrosch** [comment]
If you want to manually tweak our unit file, do so by `systemctl edit` ...

**Comment by geotekberlin** [comment]
TasksMax=infinity in the icinga2.service unit file does not work either with systemd 228, this had worked with earlier versions though

**Comment by lazyfrosch** [comment]
@geotekberlin could this be to other defaults set in OpenSUSE lately?

**Comment by lazyfrosch** [comment]
@geotekberlin How big is your environment?

Could you also try to raise `LimitNPROC=62883` for the unit. (Don't forget `systemctl daemon-reload`)

**Comment by geotekberlin** [comment]
We are checking about 400 hosts and 3500 services on one master and one satellite instances. About 100 hosts are windows based an are running the Icinga Agent. 

Icinga is rock stable since we set DefaultTasksMax in system.conf. What exactly do you want me to check by also setting the LimitNPROC setting?

**Comment by lazyfrosch** [comment]
We suspect that `LimitNPROC` might have an effect on big setups, since it limits some of the process resources.

`TasksMax` should be set correctly in the next release. We are discussing backwards compatibility with older systemd versions.

**Comment by geotekberlin** [comment]
It took a while to set up a mirrored icinga instance because we can't risk our production system to fail.

After setting `LimitNPROC=62883` and removing `TasksMax` in the icinga2 unit file and without a `DefaultTasksMax=infinity` setting in `/etc/systemd/system.conf`  the mirror system crashed after running for 30 hours. So this does not seem to work.

The really strange thing is that setting TasksMax=infinity in the icinga2.service unit file has no effect any more with the current OpenSuse 42.3 release build / Icinga2 2.7.1 combination. `systemctl show -p TasksMax Icinga2` still shows `TaskMax=512`, even after a complete reboot. This did work though an an earlier OpenSuse 42.3 release build with Icinga2. The only thing that had changed was the systemd version, which had been upped by the OpenSuse automatic updates.

**Comment by dnsmichi** [comment]
Am I correct with the assumption that your Icinga 2 master runs on OpenSuse 42.3, and this is not a client? I would turn off automatic updates and control that on my own if I were you. Are you able to downgrade systemd?

I haven't found any bugs with systemd and this setting, so this should work. Maybe the implementation inside the kernel, or systemd is buggy.

Other projects affected from TasksMax:

moby/moby#24197
https://bugs.launchpad.net/charms/+source/percona-cluster/+bug/1578080
http://blog.nashcom.de/nashcomblog.nsf/dx/sles-12-sp2-issues-with-domino-running-with-systemd.htm
http://tracker.ceph.com/issues/15583
https://bugzilla.redhat.com/show_bug.cgi?id=1337244
https://bugs.archlinux.org/task/47303
https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/YBYAAME2YGXQ3EVKMOQPB7D4CUYFF4YE/
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=823530
https://www.elastic.co/blog/we-are-out-of-memory-systemd-process-limits

Upstream discussion: systemd/systemd#3211

**Comment by geotekberlin** [comment]
You are correct that Icinga2 master runs on OpenSuse 42.3. We are administering all sorts of OpenSuse VMs for several years and had **never ever** seen any problems that were related to Suse bugs, so I am very happy with Suse's code quality. 

I am aware that the Icinga crashes are caused by systemd changes and you are correctly citing several other projects that were also hit by these changes of behaviour.

However, the observed effect that TasksMax has no effect any more may well be an OpenSuse 42.3 bug, but I have no time to investigate it further. At least not as long as the DefaulTasksMax setting is a viable workaround.

**Comment by lazyfrosch** [comment]
This will also be fixed by packaging with a icinga2-2.7.1-2

Released for:
* Debian stretch
* Ubuntu >= xenial
* OpenSUSE >= 43.2

See:
* https://github.com/Icinga/icinga-packaging/pull/48
* https://github.com/Icinga/icinga-packaging/pull/49
